### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by tyson-marks29**

### CODE REVIEW

File: add_book.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
+<?php
+include 'database.php';
+
+if ($_SERVER['REQUEST_METHOD'] == 'POST') {
```
Comment (tyson-marks29) [inline add_book.php:4]: Consider using prepared statements to prevent SQL injection attacks. Directly using user input in SQL queries is a security risk. PHP's `mysqli` library supports prepared statements which can mitigate this risk.

```php
$stmt = $connection->prepare("INSERT INTO books (title, author, year) VALUES (?, ?, ?)");
$stmt->bind_param("ssi", $title, $author, $year);
if ($stmt->execute()) {
    echo "Book added successfully!";
} else {
    echo "Error: " . $stmt->error;
}
$stmt->close();
```

Comment (tyson-marks29) [inline add_book.php:4]: The current implementation lacks validation for the POST data. It's important to validate and sanitize user inputs to ensure data integrity and prevent potential security issues.

```php
$title = filter_input(INPUT_POST, 'title', FILTER_SANITIZE_STRING);
$author = filter_input(INPUT_POST, 'author', FILTER_SANITIZE_STRING);
$year = filter_input(INPUT_POST, 'year', FILTER_VALIDATE_INT);

if (!$title || !$author || !$year) {
    echo "Invalid input data.";
    exit;
}
```

Comment (tyson-marks29) [inline add_book.php:4]: The database connection is being included directly in the script, which may not be optimal for larger applications. Consider using a Service or Repository pattern to handle database operations, which will improve maintainability and testability.

Create a `BookRepository` class:

```php
class BookRepository {
    private $connection;

    public function __construct($connection) {
        $this->connection = $connection;
    }

    public function addBook($title, $author, $year) {
        $stmt = $this->connection->prepare("INSERT INTO books (title, author, year) VALUES (?, ?, ?)");
        $stmt->bind_param("ssi", $title, $author, $year);
        return $stmt->execute();
    }
}
```

Then use this class in your script:

```php
$bookRepository = new BookRepository($connection);
if ($bookRepository->addBook($title, $author, $year)) {
    echo "Book added successfully!";
} else {
    echo "Error: " . $connection->error;
}
```

Comment (tyson-marks29) [inline add_book.php:4]: This code does not currently have any unit tests. Consider adding tests to ensure that the logic for adding a book is functioning correctly. You can use PHPUnit for this purpose.

Example test:

```php
use PHPUnit\Framework\TestCase;

class BookRepositoryTest extends TestCase {
    public function testAddBook() {
        $mockConnection = $this->createMock(mysqli::class);
        $mockStmt = $this->createMock(mysqli_stmt::class);

        $mockConnection->method('prepare')->willReturn($mockStmt);
        $mockStmt->expects($this->once())->method('bind_param')->with($this->equalTo('ssi'), $this->anything(), $this->anything(), $this->anything());
        $mockStmt->method('execute')->willReturn(true);

        $repository = new BookRepository($mockConnection);
        $this->assertTrue($repository->addBook('Test Title', 'Test Author', 2021));
    }
}
```

File: database.php [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?php
+$connection = mysqli_connect("localhost", "root", "", "books_db");
+
+if (!$connection) {
```
Comment (tyson-marks29) [inline database.php:4]: There's no error handling for the database connection. Consider using exceptions to handle connection errors more gracefully and provide more detailed error reporting.

```php
try {
    $connection = new mysqli("localhost", "root", "", "books_db");
    if ($connection->connect_error) {
        throw new Exception("Connection failed: " . $connection->connect_error);
    }
} catch (Exception $e) {
    echo "Error: " . $e->getMessage();
    exit;
}
```

File: delete_book.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<?php
+include 'database.php';
+
+if (isset($_GET['id'])) {
```
Comment (tyson-marks29) [inline delete_book.php:4]: The code for deleting a book is vulnerable to SQL injection. Use prepared statements to mitigate this risk.

```php
if (isset($_GET['id'])) {
    $id = intval($_GET['id']); // Ensure ID is an integer
    $stmt = $connection->prepare("DELETE FROM books WHERE id = ?");
    $stmt->bind_param("i", $id);
    if ($stmt->execute()) {
        echo "Book deleted successfully!";
        header("Location: list_books.php");
    } else {
        echo "Error: " . $stmt->error;
    }
    $stmt->close();
}
```

Comment (tyson-marks29) [inline delete_book.php:4]: Consider refactoring the database operations into a repository class for better separation of concerns and maintainability.

```php
class BookRepository {
    private $connection;

    public function __construct($connection) {
        $this->connection = $connection;
    }

    public function deleteBook($id) {
        $stmt = $this->connection->prepare("DELETE FROM books WHERE id = ?");
        $stmt->bind_param("i", $id);
        return $stmt->execute();
    }
}
```

Usage:

```php
$bookRepository = new BookRepository($connection);
if ($bookRepository->deleteBook($id)) {
    echo "Book deleted successfully!";
    header("Location: list_books.php");
} else {
    echo "Error: " . $connection->error;
}
```

Comment (tyson-marks29) [inline delete_book.php:4]: The current implementation does not include any unit tests for the delete functionality. Consider creating tests to verify that books are deleted correctly and handle cases where deletion might fail.

```php
use PHPUnit\Framework\TestCase;

class BookRepositoryTest extends TestCase {
    public function testDeleteBook() {
        $mockConnection = $this->createMock(mysqli::class);
        $mockStmt = $this->createMock(mysqli_stmt::class);

        $mockConnection->method('prepare')->willReturn($mockStmt);
        $mockStmt->expects($this->once())->method('bind_param')->with($this->equalTo('i'), $this->anything());
        $mockStmt->method('execute')->willReturn(true);

        $repository = new BookRepository($mockConnection);
        $this->assertTrue($repository->deleteBook(1));
    }
}
```

File: edit_book.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<?php
+include 'database.php';
+
+if (isset($_GET['id'])) {
```
Comment (tyson-marks29) [inline edit_book.php:4]: The edit functionality is also vulnerable to SQL injection. Use prepared statements to protect against this.

```php
if ($_SERVER['REQUEST_METHOD'] == 'POST') {
    $id = intval($_POST['id']);
    $title = $_POST['title'];
    $author = $_POST['author'];
    $year = intval($_POST['year']);

    $stmt = $connection->prepare("UPDATE books SET title = ?, author = ?, year = ? WHERE id = ?");
    $stmt->bind_param("ssii", $title, $author, $year, $id);
    if ($stmt->execute()) {
        echo "Book updated successfully!";
        header("Location: list_books.php");
    } else {
        echo "Error: " . $stmt->error;
    }
    $stmt->close();
}
```

Comment (tyson-marks29) [inline edit_book.php:4]: Refactor the database operations into a repository class for better separation of concerns and maintainability.

```php
class BookRepository {
    private $connection;

    public function __construct($connection) {
        $this->connection = $connection;
    }

    public function updateBook($id, $title, $author, $year) {
        $stmt = $this->connection->prepare("UPDATE books SET title = ?, author = ?, year = ? WHERE id = ?");
        $stmt->bind_param("ssii", $title, $author, $year, $id);
        return $stmt->execute();
    }
}
```

Usage:

```php
$bookRepository = new BookRepository($connection);
if ($bookRepository->updateBook($id, $title, $author, $year)) {
    echo "Book updated successfully!";
    header("Location: list_books.php");
} else {
    echo "Error: " . $connection->error;
}
```

Comment (tyson-marks29) [inline edit_book.php:4]: Consider adding unit tests for the update functionality. This will help ensure that the edit logic works as expected and can handle potential edge cases.

```php
use PHPUnit\Framework\TestCase;

class BookRepositoryTest extends TestCase {
    public function testUpdateBook() {
        $mockConnection = $this->createMock(mysqli::class);
        $mockStmt = $this->createMock(mysqli_stmt::class);

        $mockConnection->method('prepare')->willReturn($mockStmt);
        $mockStmt->expects($this->once())->method('bind_param')->with($this->equalTo('ssii'), $this->anything(), $this->anything(), $this->anything(), $this->anything());
        $mockStmt->method('execute')->willReturn(true);

        $repository = new BookRepository($mockConnection);
        $this->assertTrue($repository->updateBook(1, 'New Title', 'New Author', 2022));
    }
}
```

File: list_books.php [hunk 5]
```diff
@@ -0,0 +1,34 @@
+<?php
+include 'database.php';
+
+$query = "SELECT * FROM books";
```
Comment (tyson-marks29) [inline list_books.php:4]: The list_books.php file also directly executes SQL queries without using prepared statements. Although this is less risky since it doesn't involve user input, it's a good practice to consistently use prepared statements for all database interactions.

Consider refactoring the query execution into a repository pattern as shown in previous examples.

Comment (tyson-marks29) [inline list_books.php:4]: The current implementation lacks unit tests for listing books. While this functionality is straightforward, consider adding tests to ensure that the expected data is retrieved from the database.

```php
use PHPUnit\Framework\TestCase;

class BookRepositoryTest extends TestCase {
    public function testListBooks() {
        $mockConnection = $this->createMock(mysqli::class);
        $mockResult = $this->createMock(mysqli_result::class);

        $mockConnection->method('query')->willReturn($mockResult);
        $mockResult->method('fetch_assoc')->willReturnOnConsecutiveCalls(
            ['title' => 'Book 1', 'author' => 'Author 1', 'year' => 2020],
            null
        );

        $repository = new BookRepository($mockConnection);
        $books = $repository->listBooks();

        $this->assertEquals('Book 1', $books[0]['title']);
        $this->assertEquals('Author 1', $books[0]['author']);
        $this->assertEquals(2020, $books[0]['year']);
    }
}
```
